lib/puma/minissl.rb

```diff
@@ -40,7 +91,18 @@ def readpartial(size)
       end
 
       def engine_read_all
-        output = @engine.read
+        begin
+          output = @engine.read
+          return output if bad_tlsv1_3?
+        rescue Puma::MiniSSL::SSLError => e
+          if e.message.end_with? 'System error: Undefined error: 0 - 0'
```

Comment:
This doesn't look robust to me, matching against the message... do we know where does this error come from? What does `Undefined error: 0 - 0` mean?

At https://github.com/puma/puma/issues/1708#issuecomment-552079685 I saw the similar but different error

    Error in reactor loop escaped: System error: Success - 0 (Puma::MiniSSL::SSLError)

Reply:
@dentarg

> matching against the message

I wasn't happy with using a string match either.  Let me get back into this and see what can be done.

> What does Undefined error: 0 - 0 mean?

Similar to what happened with 1.1.1e.  The error previously had a numeric code of 0, then with 1.1.1e they changed the code, which was a breaking change.  I'll look at the code for `Puma::MiniSSL::SSLError` and see if better info can be passed from the c code.

This issue is messy, as flooding an https server with http connections could be used as an DOS attack.

We want Puma to reject http connections, and we want that done as quickly as possible so a minimal amount of resources are used by them.  We also don't want to slow down real https connections.

The issue is made worse by the fact that TLSv1.3 and TLSv1.2 handle http connections differently, and right now, I can't recall that exact difference.  As I recall, TLSv1.2 handled the problem itself, TLSv1.3 forces Puma to deal with it...